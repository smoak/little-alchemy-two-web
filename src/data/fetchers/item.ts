const database = import('../db.json');

// type DatabaseSchema = {
//   readonly items: DatabaseItems;
// };

export type DatabaseItem = {
  readonly myths: boolean;
  readonly makes: { source: string; target: string }[];
  readonly combinations: string[][];
  readonly image_url: string;
};

type DatabaseItems = Record<string, DatabaseItem>;

type GetItems = () => Promise<DatabaseItems>;
export const getItems: GetItems = async () => (await database).items;

type GetItemByName = (name: string) => Promise<DatabaseItem | undefined>;
export const getItemByName: GetItemByName = async (name) => {
  const items = await getItems();
  return items[name];
};

export const getItemNames = async () => {
  const items = await getItems();
  return Object.keys(items);
};
