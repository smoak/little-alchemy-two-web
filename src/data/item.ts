export const itemDisplayName = (itemServerName: string) => itemServerName.replaceAll('-', ' ');
