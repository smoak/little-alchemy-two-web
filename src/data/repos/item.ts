import Fuse, { IFuseOptions } from 'fuse.js';

import { DatabaseItem, getItemByName, getItemNames, getItems } from '../fetchers/item';
import type { Item, ItemCombination } from '../types';

const fromCombinations = (combinations: DatabaseItem['combinations']): ItemCombination[] =>
  combinations.map((combo) => ({ source: combo[0], target: combo[1] }));

const fromItemWithName = (name: string, item: DatabaseItem): Item => ({
  name,
  imageUrl: item.image_url,
  myths: item.myths,
  creates: item.makes,
  combinations: fromCombinations(item.combinations),
});

type FindById = (id: string) => Promise<Item>;
export const findById: FindById = async (id) => {
  const item = await getItemByName(id);

  if (!item) {
    throw new Error(`Could not find item with id: ${id}`);
  }

  return fromItemWithName(id, item);
};

type FindAll = () => Promise<Item[]>;
export const findAll: FindAll = async () => {
  const items = await getItems();

  return Object.keys(items).map<Item>((itemName) => {
    const item = items[itemName];
    return fromItemWithName(itemName, item);
  });
};

type Search = (query: string) => Promise<Item[]>;
export const search: Search = async (query) => {
  const items = await getItemNames();
  const options: IFuseOptions<string> = {
    minMatchCharLength: 2,
    threshold: 1,
  };
  const fuse = new Fuse(items, options);
  const results = fuse.search(query);
  return Promise.all(results.map((result) => findById(result.item)));
};
