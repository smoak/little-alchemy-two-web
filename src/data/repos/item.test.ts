jest.mock('../fetchers/item');

import { DatabaseItem, getItemByName, getItemNames, getItems } from '../fetchers/item';

import { findAll, findById, search } from './item';

describe('item', () => {
  const item: DatabaseItem = {
    combinations: [['item1', 'item2']],
    makes: [
      {
        source: 'item1',
        target: 'item2',
      },
    ],
    myths: false,
    image_url: 'https://picsum.photos/48',
  };

  describe('.findById', () => {
    let result: ReturnType<typeof findById>;

    describe('when the item cannot be found', () => {
      beforeEach(() => {
        (getItemByName as jest.Mock).mockReturnValue(undefined);
      });

      it('throws an error', () => {
        expect(() => findById('foo')).toThrowErrorMatchingSnapshot();
      });
    });

    describe('when the item can be found', () => {
      beforeEach(() => {
        (getItemByName as jest.Mock).mockReturnValue(item);

        result = findById('item');
      });

      it('returns the expected result', () => {
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('.findAll', () => {
    let result: ReturnType<typeof findAll>;

    describe('when there are no items', () => {
      beforeEach(() => {
        (getItems as jest.Mock).mockReturnValue([]);

        result = findAll();
      });

      it('retuns the expected result', () => {
        expect(result).toEqual([]);
      });
    });

    describe('when there are items', () => {
      beforeEach(() => {
        const items = {
          ['item']: item,
        };
        (getItems as jest.Mock).mockReturnValue(items);

        result = findAll();
      });

      it('returns the expected result', () => {
        expect(result).toMatchSnapshot();
      });
    });
  });

  describe('.search', () => {
    let result: ReturnType<typeof search>;

    describe('when the query doesnt match anything', () => {
      beforeEach(() => {
        (getItemNames as jest.Mock).mockReturnValue(['item1']);

        result = search('nomatch');
      });

      it('returns the expected result', () => {
        expect(result).toMatchSnapshot();
      });
    });

    describe('when the query matches', () => {
      beforeEach(() => {
        (getItemNames as jest.Mock).mockReturnValue(['item1']);
        (getItemByName as jest.Mock).mockReturnValue(item);

        result = search('item');
      });

      it('returns the expected result', () => {
        expect(result).toMatchSnapshot();
      });
    });
  });
});
