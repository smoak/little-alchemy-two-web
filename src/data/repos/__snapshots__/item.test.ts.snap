// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`item .findAll when there are items returns the expected result 1`] = `
[
  {
    "combinations": [
      {
        "source": "item1",
        "target": "item2",
      },
    ],
    "creates": [
      {
        "source": "item1",
        "target": "item2",
      },
    ],
    "imageUrl": "https://picsum.photos/48",
    "myths": false,
    "name": "item",
  },
]
`;

exports[`item .findById when the item can be found returns the expected result 1`] = `
{
  "combinations": [
    {
      "source": "item1",
      "target": "item2",
    },
  ],
  "creates": [
    {
      "source": "item1",
      "target": "item2",
    },
  ],
  "imageUrl": "https://picsum.photos/48",
  "myths": false,
  "name": "item",
}
`;

exports[`item .findById when the item cannot be found throws an error 1`] = `"Could not find item with id: foo"`;

exports[`item .search when the query doesnt match anything returns the expected result 1`] = `[]`;

exports[`item .search when the query matches returns the expected result 1`] = `
[
  {
    "combinations": [
      {
        "source": "item1",
        "target": "item2",
      },
    ],
    "creates": [
      {
        "source": "item1",
        "target": "item2",
      },
    ],
    "imageUrl": "https://picsum.photos/48",
    "myths": false,
    "name": "item1",
  },
]
`;
